File: leetcode/p0847/Solution.java
class Solution {
    public int shortestPathLength(int[][] graph) {
        Queue<List<Integer>> q = new ArrayDeque<>(); // state node cost
        Set<List<Integer>> seen = new HashSet<>(); // state node
        
        for(int i = 0; i < graph.length; ++i) {
            q.offer(Arrays.asList((1<<i), i, 0));
            seen.add(Arrays.asList((1<<i), i));
        }
        final int finalState = (1 << graph.length) - 1;
        while(!q.isEmpty()) {
            List<Integer> item = q.poll();
            int state = item.get(0);
            int node = item.get(1);
            int cost = item.get(2);
            
            if (state == finalState) {
                return cost;
            }
            
            for(int next : graph[node]) {
                int nextState = state | (1<<next);
                if (seen.contains(Arrays.asList(nextState,next))) {
                    continue;
                }
                q.offer(Arrays.asList(nextState, next, cost + 1));
                seen.add(Arrays.asList(nextState, next));
            }
        }
        return 0;
    }
}

File: leetcode/p0847/Solution.kt
class Solution {
    fun shortestPathLength(graph: Array<IntArray>): Int {
        val q = ArrayDeque<Triple<Int,Int,Int>>() // state node cost
        val seen = mutableSetOf<Pair<Int,Int>>() // state node
        
        for(i in 0 until graph.size) {
            q.offer(Triple((1 shl i),i,0))
            seen.add(Pair((1 shl i),i))
        }
        
        val finalState = (1 shl graph.size) - 1
        while(!q.isEmpty()) {
            val (state, node, cost) = q.poll()
            if (state == finalState) {
                return cost
            }
            
            for (next in graph[node]) {
                val nextState = state or (1 shl next)
                if (seen.contains(Pair(nextState, next))) {
                    continue
                }
                q.offer(Triple(nextState,next, cost+1))
                seen.add(Pair(nextState, next))
            }
        }
        return 0
    }
} 
